components/encryption/src/manager/mod.rs

```diff
@@ -786,16 +843,168 @@ impl EncryptionKeyManager for DataKeyManager {
         fail_point!("key_manager_fails_before_delete_file", |_| IoResult::Err(
             std::io::ErrorKind::Other.into()
         ));
-        self.dicts.delete_file(fname)?;
+        // `RemoveDir` is not managed, but RocksDB may use `RenameFile` on a directory,
+        // which internally calls `LinkFile` and `DeleteFile`.
+        let path = Path::new(fname);
+        if path.is_dir() {
+            let mut iter = walkdir::WalkDir::new(path).into_iter().peekable();
+            while let Some(e) = iter.next() {
+                self.dicts.delete_file(
+                    e?.path().as_os_str().to_str().unwrap(),
+                    iter.peek().is_none(),
+                )?;
+            }
+        } else {
+            self.dicts.delete_file(fname, true)?;
+        }
         Ok(())
     }
 
     fn link_file(&self, src_fname: &str, dst_fname: &str) -> IoResult<()> {
-        self.dicts.link_file(src_fname, dst_fname)?;
+        let src_path = Path::new(src_fname);
+        let dst_path = Path::new(dst_fname);
+        if src_path.is_dir() {
+            let mut iter = walkdir::WalkDir::new(src_path).into_iter().peekable();
+            while let Some(e) = iter.next() {
+                let e = e?;
+                if e.path_is_symlink() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("unexpected symbolic link: {}", e.path().display()),
+                    ));
+                }
+                if e.path().is_dir() {
+                    continue;
+                }
+                let sub_path = e.path().strip_prefix(src_path).unwrap();
+                let src = e.path().as_os_str().to_str().unwrap();
+                let dst_path = dst_path.join(sub_path);
+                let dst = dst_path.as_os_str().to_str().unwrap();
+                self.dicts.link_file(src, dst, iter.peek().is_none())?;
```

Comment:
what if the last item is a directory?

Reply:
Directories are ignored. Walkdir traverses the dir recursively, so all files will be handled even though subdir is skipped.

Reply:
Is there's a possibility that the last item is an empty dir so the last iteration will be skipped at line877 and no sync is ever passed  to `link_file`?

Reply:
Fixed.